fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,779 @@
+    * @return double maximum characteristic speed, |u| + √(γp/ρ)
+    */
+   double ComputeFluxDotN(const Vector &x, const Vector &normal,
+                          ElementTransformation &Tr, Vector &FUdotN)
+   {
+      const int dim = normal.Size();
+
+      // 1. Get states
+      const double density = x(0);                  // ρ
+      const Vector momentum(x.GetData() + 1, dim);  // ρu
+      const double energy = x(1 + dim);             // E, internal energy ρe
+      // pressure, p = (γ-1)*(E - ½ρ|u|^2)
+      const double pressure = (specific_heat_ratio - 1.0) *
+                              (energy - 0.5 * (momentum * momentum) / density);
+
+      // Check whether the solution is physical only in debug mode
+      MFEM_ASSERT(density >= 0, "Negative Density");
+      MFEM_ASSERT(pressure >= 0, "Negative Pressure");
+      MFEM_ASSERT(energy >= 0, "Negative Energy");
+
+      // 2. Compute normal flux
+
+      FUdotN(0) = momentum * normal;  // ρu⋅n
+      // u⋅n
+      const double normal_velocity = FUdotN(0) / density;
+      for (int d = 0; d < dim; d++)
+      {
+         // (ρuuᵀ + pI)n = ρu*(u⋅n) + pn
+         FUdotN(1 + d) = normal_velocity * momentum(d) + pressure * normal(d);
+      }
+      // (u⋅n)(E + p)
+      FUdotN(1 + dim) = normal_velocity * (energy + pressure);
+
+      // 3. Compute maximum characteristic speed
+
+      // sound speed, √(γ p / ρ)
+      const double sound = sqrt(specific_heat_ratio * pressure / density);
+      // fluid speed |u|
+      const double speed = sqrt(momentum * momentum) / density;
+      // max characteristic speed = fluid speed + sound speed
+      return speed + sound;
+   }
+
+   /**
+    * @brief Construct a new Euler Element Form Integrator object with given
+    * integral order offset
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param dim spatial dimension
+    * @param specific_heat_ratio_ specific heat ratio, γ
+    * @param IntOrderOffset_ 2*p + IntOrderOffset will be used for quadrature
+    */
+   EulerFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                       const double specific_heat_ratio_,
+                       const int IntOrderOffset_)
+      : HyperbolicFormIntegrator(rsolver_, dim, dim + 2, IntOrderOffset_),
+        specific_heat_ratio(specific_heat_ratio_) {}
+
+   /**
+    * @brief Construct a new Euler Element Form Integrator object with given
+    * integral rule
+    *
+    * @param[in] rsolver_ numerical flux
+    * @param dim spatial dimension
+    * @param specific_heat_ratio_ specific heat ratio, γ
+    * @param ir this integral rule will be used for the Gauss quadrature
+    */
+   EulerFormIntegrator(RiemannSolver *rsolver_, const int dim,
+                       const double specific_heat_ratio_,
+                       const IntegrationRule *ir)
+      : HyperbolicFormIntegrator(rsolver_, dim, dim + 2, ir),
+        specific_heat_ratio(specific_heat_ratio_) {}
+};
+}
+
+#endif
```

Comment:
Newline at end of file
```suggestion
#endif

```